fastchat/model/compression.py

```diff
@@ -117,15 +136,39 @@ def load_compress_model(model_path, device, torch_dtype, use_fast, revision="mai
             model_path,
             low_cpu_mem_usage=True,
             torch_dtype=torch_dtype,
-            trust_remote_code=True,
             revision=revision,
+            trust_remote_code=True,
         )
         # some models are loaded by AutoModel but not AutoModelForCausalLM,
         # such as chatglm, chatglm2
         try:
-            model = AutoModelForCausalLM.from_config(config, trust_remote_code=True)
+            model = AutoModelForCausalLM.from_config(
+                config, trust_remote_code=True, device=device
+            )
```

Comment:
I just remembered why you cannot use `device=device` here. This is because we always want to initialize the model on the CPU.

For example, if a model is 20GB and your GPU is 16GB. You should be able to run the model with 4-bit compression if you do compression before loading it to GPU, which is the code before your modification.
However, after your modification, it will hit OOM in this initialization line, making it unable to run this model.

Please correct me if I am wrong. Otherwise, please revert your change and make sure your code initializes the model on the CPU, then compress it and load it to the GPU layer-by-layer.

Reply:
I suppose I misunderstand you. Do you mean the code in AutoModel.from_config or AutoModelForCausalLM.from_config?

Reply:
yes